src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -183,51 +183,80 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
         continue;
       }
 
-      std::shared_ptr<WorkerInterface> worker = worker_pool_.PopWorker(spec);
-      if (!worker) {
-        // No worker available, we won't be able to schedule any kind of task.
-        return;
-      }
-
       const auto owner_worker_id = WorkerID::FromBinary(spec.CallerAddress().worker_id());
       const auto owner_node_id = NodeID::FromBinary(spec.CallerAddress().raylet_id());
+
       // If the owner has died since this task was queued, cancel the task by
       // killing the worker (unless this task is for a detached actor).
       if (!spec.IsDetachedActor() && !is_owner_alive_(owner_worker_id, owner_node_id)) {
         RAY_LOG(WARNING) << "Task: " << task.GetTaskSpecification().TaskId()
                          << "'s caller is no longer running. Cancelling task.";
-        worker_pool_.PushWorker(worker);
         if (!spec.GetDependencies().empty()) {
           task_dependency_manager_.RemoveTaskDependencies(
               task.GetTaskSpecification().TaskId());
         }
         work_it = dispatch_queue.erase(work_it);
+        continue;
+      }
+
+      // Check if the node is still schedulable. It may not be if dependency resolution
+      // took a long time.
+      std::shared_ptr<TaskResourceInstances> allocated_instances(
+          new TaskResourceInstances());
+      bool schedulable = cluster_resource_scheduler_->AllocateLocalTaskResources(
+          spec.GetRequiredResources().GetResourceMap(), allocated_instances);
+
+      if (!schedulable) {
+        // The local node currently does not have the resources to run the task, so we
+        // should try spilling to another node.
+        bool did_spill = TrySpillback(work);
+        if (!did_spill) {
+          // This must be the only feasible node in the cluster, so keep the task on the
+          // dispatch queue.
+          break;
+        }
       } else {
-        bool worker_leased;
-        bool remove = AttemptDispatchWork(*work_it, worker, &worker_leased);
-        if (worker_leased) {
-          // Pin the arguments while the lease is active. These will be erased
-          // once the lease is returned.
-          num_pinned_task_arguments_ += args.size();
-          RAY_CHECK(pinned_task_arguments_.emplace(spec.TaskId(), std::move(args)).second)
-              << spec.TaskId();
-
-          auto reply = std::get<1>(*work_it);
-          auto callback = std::get<2>(*work_it);
-          Dispatch(worker, leased_workers_, task, reply, callback);
-        } else {
-          worker_pool_.PushWorker(worker);
+        // The local node has the available resources to run the task, so we should run
```

Comment:
It's getting a bit hard to read; could you split this block of code out into a helper function like `TrySpillback`?

Reply:
Yeah, we tried moving it into a helper, but it depends on a bunch of random things so it won't be as clean. I can move it as an organizational thing I guess